[libc++][doc] Use installed std modules in external projects. (#80601)
llvm/llvm-project: libcxx/docs/Modules.rst

   system vendors, with the goal that building the BMI files is done by
   the build system.

> Currently this requires a local build of libc++ with modules installation enabled.
> Since modules are not installed by default. You can build and install the modules
> to ``<install_prefix>`` with the following commands.

Reviewer: ```suggestion
This requires the system you're distributes the installed std modules. Otherwise...
```

Author: Thanks for your review. But I do not understand why you want me to delete these three lines. Could you please explain more in detail?
And also, The line you suggest me to add is not complete and seems to have a grammer error. Could you please tell me what you want to express?

Reviewer: Sorry. I mean it may be better to not say "the modules are not installed by default". Since there may be system vendors enable it.

Author: Ok, got it. So you mean "requires the system you're using distributes ..." right? And "Otherwise, you need to build and install ..." right? I think this suggestion is fine. But I also want to hear from main author of the original page @mordante.

Reviewer 2: We shouldn't change this in this fashion. The current described way for building is broken, this should be fixed. So that way remains a viable option. The goals of this patch is to add a second option based on installed modules.

.. code-block:: bash

>   $ git clone https://github.com/llvm/llvm-project.git --depth 1

Reviewer 2: I prefer to keep the instructions generic, if users like `--depth 1` they are free to do so.

Author: OK

  $ cd llvm-project
  $ mkdir build
  $ cmake -G Ninja -S runtimes -B build -DCMAKE_C_COMPILER=<path-to-compiler> -DCMAKE_CXX_COMPILER=<path-to-compiler> -DLIBCXX_INSTALL_MODULES=ON -DLLVM_ENABLE_RUNTIMES="libcxx;libcxxabi;libunwind"
  $ cmake --build build -- -j $(nproc)
  $ cmake --install build --prefix <install_prefix>

This is a small sample program that uses the module ``std``. It consists of a
``CMakeLists.txt``, an ``std.cmake``, and a ``main.cpp`` file.

.. code-block:: cpp

  // main.cpp
  import std; // When importing std.compat it's not needed to import std.
  import std.compat;

  int main() {

>     std::println("Hello modular world");

Reviewer 2: please don't make unrelated changes in the same patch.

Author: OK, will change it back

Currently this requires a local build of libc++ with modules enabled. Since
modules are not part of the installation yet, they are used from the build
directory. First libc++ needs to be build with module support enabled. If
the libc++ is built with installation of modules enabled, they can also be
used from the installation directory. The following instructions enable the
installation of modules and install the modules into ``<install_prefix>``.

.. code-block:: bash

  $ git clone https://github.com/llvm/llvm-project.git --depth 1
  $ cd llvm-project
  $ mkdir build
  $ cmake -G Ninja -S runtimes -B build -DLIBCXX_INSTALL_MODULES=ON -DLLVM_ENABLE_RUNTIMES="libcxx;libcxxabi;libunwind"
  $ cmake --build build -- -j $(nproc)

>   $ cmake --install build --prefix <install_prefix>

Reviewer 2: I really prefer to keep the `ninja -C build` and add `ninja install -C build`. This is the style we use in other parts of the libc++ documentation.

  #
  # Import the modules from libc++
  #
  include(std.cmake)

  add_executable(main main.cpp)

.. code-block:: cmake

  # std.cmake
  include(FetchContent)
  FetchContent_Declare(
    std_module
    URL "file://${LIBCXX_INSTALLED_DIR}/share/libc++/v1"
    DOWNLOAD_EXTRACT_TIMESTAMP TRUE
    SYSTEM
  )

  if (NOT std_module_POPULATED)
    FetchContent_Populate(std_module)
  endif()

  #
  # Add std static library
  #

>   add_library(std)

Reviewer 3: Missing "STATIC"?

Author: I will add it

  #
  # Import the modules from libc++
  #
  include(std.cmake)

  add_executable(main main.cpp)

.. code-block:: cmake

  # std.cmake
  include(FetchContent)
  FetchContent_Declare(
    std_module

>     URL "file://${LIBCXX_INSTALLED_DIR}/share/libc++/v1"

Reviewer 2: This part of the example should use the `${LIBCXX_BUILD}` right? Are the other parts of the changes in this part intended?

.. note:: The ``std`` dependencies of ``std.compat`` is not always resolved when
          building the ``std`` target using multiple jobs.

This is another small sample program that uses the module ``std`` from

> installation directory. It consists of a ``CMakeLists.txt``, an
> ``std.cmake``, and a ``main.cpp`` file. The ``main.cpp`` is the same as
> the previous example.

Reviewer 2: ```suggestion
the installation directory. It consists of a ``CMakeLists.txt``,
a ``std.cmake``, and the ``main.cpp`` from the previous example.
```

      -Wno-reserved-user-defined-literal
  )

  target_compile_options(std
    PUBLIC
      -nostdinc++
  )

  #
  # Adjust project linker flags
  #

  target_link_options(std
    INTERFACE
      -nostdlib++
      -L${LIBCXX_INSTALLED_DIR}/lib
      -Wl,-rpath,${LIBCXX_INSTALLED_DIR}/lib
  )

  target_link_libraries(std
    INTERFACE
      c++
  )
  
  #
  # Link to the std modules by default
  #

  link_libraries(std)

Building this project is done with the following steps, assuming the files
``main.cpp``, ``CMakeLists.txt``, and ``std.cmake`` are copied in the current directory.

.. code-block:: bash

  $ mkdir build
  $ cmake -S . -B build -G Ninja -DCMAKE_CXX_COMPILER=<path-to-compiler> -DLIBCXX_INSTALLED_DIR=<install_prefix>
  $ ninja -C build
  $ build/main

.. warning:: You need more than clang itself to build a project using modules.
             Specifically, you will need ``clang-scan-deps``. For example, in Ubuntu, you
             need to use ``sudo ./llvm.sh 17 all`` rather than ``sudo ./llvm.sh 17`` showed
             in `LLVM Debian/Ubuntu nightly packages <https://apt.llvm.org>`__ to install
             essential components to build this project.

> 

Reviewer 2: I would remove this part. This is also required when using the build directory.
